lib/zip.js

```diff
@@ -0,0 +1,379 @@
+  0x9c0906a9, 0xeb0e363f, 0x72076785, 0x05005713,
+  0x95bf4a82, 0xe2b87a14, 0x7bb12bae, 0x0cb61b38,
+  0x92d28e9b, 0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21,
+  0x86d3d2d4, 0xf1d4e242, 0x68ddb3f8, 0x1fda836e,
+  0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777,
+  0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c,
+  0x8f659eff, 0xf862ae69, 0x616bffd3, 0x166ccf45,
+  0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2,
+  0xa7672661, 0xd06016f7, 0x4969474d, 0x3e6e77db,
+  0xaed16a4a, 0xd9d65adc, 0x40df0b66, 0x37d83bf0,
+  0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
+  0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6,
+  0xbad03605, 0xcdd70693, 0x54de5729, 0x23d967bf,
+  0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
+  0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d,
+];
+
+class Stats {
+  constructor(size, compressionSize, compressionAlgorithm, mtime, crc, opsys, attributes) {
+    this.mode = opsys === ZIP_OPSYS_UNIX ? attributes >>> 16 : 0;
+    this.size = size;
+
+    this.compressionSize = compressionSize;
+    this.compressionAlgorithm = compressionAlgorithm;
+
+    // eslint-disable-next-line node-core/prefer-primordials
+    this.mtime = new Date(mtime);
+    this.mtimeMs = this.mtime.getTime();
+
+    this.opsys = opsys;
+    this.attributes = attributes;
+
+    this.crc = crc;
+  }
+
+  isDirectory() {
+    return (this.mode & S_IFMT) === S_IFDIR;
+  }
+
+  isFile() {
+    return (this.mode & S_IFMT) === S_IFREG;
+  }
+
+  isBlockDevice() {
+    return (this.mode & S_IFMT) === S_IFBLK;
+  }
+
+  isCharacterDevice() {
+    return (this.mode & S_IFMT) === S_IFCHR;
+  }
+
+  isSymbolicLink() {
+    return (this.mode & S_IFMT) === S_IFLNK;
+  }
+
+  isFIFO() {
+    return (this.mode & S_IFMT) === S_IFIFO;
+  }
+
+  isSocket() {
+    return (this.mode & S_IFMT) === S_IFSOCK;
+  }
+}
+
+class Zipent {
+  #type;
+
+  constructor(name, entry, type) {
+    this.name = name;
+    this.entry = entry;
+    this.#type = type;
+  }
+
+  isDirectory() {
+    return this.#type === S_IFDIR;
+  }
+
+  isFile() {
+    return this.#type === S_IFREG;
+  }
+
+  isBlockDevice() {
+    return this.#type === S_IFBLK;
+  }
+
+  isCharacterDevice() {
+    return this.#type === S_IFCHR;
+  }
+
+  isSymbolicLink() {
+    return this.#type === S_IFLNK;
+  }
+
+  isFIFO() {
+    return this.#type === S_IFIFO;
+  }
+
+  isSocket() {
+    return this.#type === S_IFSOCK;
+  }
+}
+
+function generateSimpleToc(entry, index) {
+  return [entry[0], index];
+}
+
+function generateZipentToc(entry, index) {
+  return [entry[0], new Zipent(entry[0], index, entry[1])];
+}
+
+class ZipArchive {
+  #zip;
+
+  constructor(data = null) {
+    this.#zip = new binding.ZipArchive(data);
+  }
+
+  getEntries({ withFileTypes = false } = {}) {
+    const toc = this.#zip.getEntries(withFileTypes);
+
+    const mapFn = withFileTypes ?
+      generateZipentToc :
+      generateSimpleToc;
+
+    // eslint-disable-next-line node-core/prefer-primordials
+    return new Map(ArrayPrototypeMap(toc, mapFn));
```

Comment:
Please use `primordials.Map` instead of the global. Passing an array to the `Map` constructor is unsafe as it relies on the user mutable `Array.prototype[Symbol.iterator]` and `%ArrayIteratorPrototype%.next`.
```suggestion
    const map = new Map();
    for(let i =0; i < toc.length; i++) MapPrototypeSet(map, mapFn(toc[i]));
    return map;
```